main.go

```diff
@@ -162,6 +164,53 @@ func dumpJSONFile(file string, pretty bool, schema bool) error {
 	return nil
 }
 
+func getFilesContentHash(files []string, tmp string) (string, bool, error) {
+	sha1 := sha1.New()
+	stdinTmpCopy := false
+
+	for i := 0; i < len(files); i++ {
+		if files[i] == "" {
+			i += 1
+			b, err := io.ReadAll(os.Stdin)
+			if err != nil {
+				return "", false, err
+			}
+
+			if _, err := sha1.Write(b); err != nil {
+				return "", false, err
+			}
+
+			file, err := os.OpenFile(tmp, os.O_RDWR, os.ModePerm)
+			if err != nil {
+				return "", false, err
+			}
+
+			if _, err := file.Write(b); err != nil {
+				return "", false, err
+			}
+			file.Close()
+			stdinTmpCopy = true
+
+		} else {
+			file, err := os.Open(files[i])
+			if err != nil {
+				return "", false, err
+			}
+			_, err = io.Copy(sha1, file)
+			if err != nil {
+				return "", false, err
+			}
+			file.Close()
```

Comment:
Ok the last thing that looks sketchy is that that the close code here and in three other places in this PR is so far from the open.

Now I know I said you can't use defer in a loop. But these loop bodies are big anyway.

So what I do in these situations is break out the body into its own function so that I *can* call defer. That way the close code is nearby the open code.

When important cleanup code is far from the initialization code you're just asking for bugs.

I'm totally ok with merging this PR wihtout this and I can do this cleanup myself because this PR has dragged on so long and I feel bad :D Or you can do it here. Either way is ok! But it should be all 3 places: line 203, line 191 and line 326

Reply:
I will clean it up, it's only a couple more lines.